tests/conftest.py

```diff
@@ -0,0 +1,125 @@
+from os import path
+import pytest
+from subprocess import (
+    check_output,
+    Popen,
+    STDOUT,
+    SubprocessError,
+    TimeoutExpired,
+)
+from time import sleep
+
+
+pytest_plugins = [
+    "mock_fixtures",
+]
+
+
+def pytest_addoption(parser):
+    parser.addoption("--nginx", default="../nginx/objs/nginx")
+    parser.addoption("--module", default="build/ngx_otel_module.so")
+    parser.addoption("--showlog", action="store_true")
+    parser.addoption("--otelcol")
+
+
+def self_signed_cert(test_dir, name):
+    k = crypto.PKey()
+    k.generate_key(crypto.TYPE_RSA, 2048)
+    cert = crypto.X509()
+    cert.get_subject().CN = name
+    cert.set_issuer(cert.get_subject())
+    cert.gmtime_adj_notBefore(0)
+    cert.gmtime_adj_notAfter(365 * 86400)  # 365 days
+    cert.set_pubkey(k)
+    cert.sign(k, "sha512")
+    (test_dir / f"{name}.key").write_text(
+        crypto.dump_privatekey(crypto.FILETYPE_PEM, k).decode("utf-8")
+    )
+    (test_dir / f"{name}.crt").write_text(
+        crypto.dump_certificate(crypto.FILETYPE_PEM, cert).decode("utf-8")
+    )
+
+
+@pytest.fixture(scope="session")
+def logger():
+    basicConfig(level=INFO)
+    return getLogger(__name__)
+
+
+@pytest.fixture(scope="module")
+def testdir(tmp_path_factory):
+    return tmp_path_factory.mktemp("nginx")
+
+
+@pytest.fixture(scope="module", autouse=True)
+def _errorlog(pytestconfig, logger, testdir):
+    yield
+    if pytestconfig.option.showlog:
+        if (testdir / "error.log").exists():
+            logger.debug((testdir / "error.log").read_text())
+
+
+@pytest.fixture(scope="module")
+def nginx_config(request, pytestconfig, testdir, logger):
+    tmpl = Environment().from_string(request.module.NGINX_CONFIG)
+    params = request.param
+    params["globals"] = (
+        f"pid {testdir}/nginx.pid;\nerror_log {testdir}/error.log debug;\n"
+        + f"load_module {path.abspath(pytestconfig.option.module)};\n"
+    )
+    params["globals_http"] = (
+        f"root {testdir};\naccess_log {testdir}/access.log;\n"
+    )
+    conf = tmpl.render(params)
+    logger.debug(conf)
+    return conf
+
+
+@pytest.fixture(scope="module")
+def nginx(testdir, pytestconfig, nginx_config, _certs, logger):
+    assert path.exists(pytestconfig.option.nginx), "No nginx binary found"
+    logger.debug(
+        check_output([pytestconfig.option.nginx, "-V"], stderr=STDOUT).decode(
+            "utf-8"
+        )
+    )
+    (testdir / "nginx.conf").write_text(nginx_config)
+    logger.info("Starting nginx...")
+    proc = Popen(
+        [
+            pytestconfig.option.nginx,
+            "-p",
+            str(testdir),
+            "-c",
+            "nginx.conf",
+            "-e",
+            "error.log",
+        ]
+    )
+    logger.debug(f"args={' '.join(proc.args)}")
+    logger.debug(f"pid={proc.pid}")
+    while not (testdir / "nginx.pid").exists():
+        sleep(0.1)
+        if proc.poll() is not None:
+            raise SubprocessError("Can't start nginx")
```

Comment:
I don't think we are supposed to throw exceptions of this type. `assert` seems more appropriate here.

Reply:
Not really, because it allows us to run further code which causes other errors. This pollutes out. Alternative could be pytest.exit() with non-0 status code.

Reply:
But raising an exception works well for me.

Reply:
`assert` throws exception...

Reply:
OK, fixed. Now assert is used.